src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexScanner.scala

```
@@ -114,7 +114,24 @@ private[oap] abstract class IndexScanner(idxMeta: IndexMeta)
 
       // Policy 1: index file size < data file size.
       val indexFileSize = indexPath.getFileSystem(conf).getContentSummary(indexPath).getLength
-      val dataFileSize = dataPath.getFileSystem(conf).getContentSummary(dataPath).getLength
+      val dataFileSize = if (dataPath.getName.endsWith(OapFileFormat.OAP_DATA_EXTENSION)) {
+        val fs = dataPath.getFileSystem(conf)
+        val fin = fs.open(dataPath)
+        val oapTotalSize = fs.getContentSummary(dataPath).getLength
+        val oapUncompressedSizeOffset = oapTotalSize - 12
+        fin.seek(oapUncompressedSizeOffset)
+        val oapUncompressedSize = fin.readLong()
+        fin.close()
+        oapUncompressedSize
+      } else {
+        // TODO: 1. Parquet file is compressed with snappy by default. It needs to figure out an
+        //          elegant approach to get the total uncompressed file size.
+        // 2. In the future, the index file may be compressed as well. At that time, this policy
+        //    will get polished or refined.
+        // 3. Below is the conservative expected compression ratio of snappy. It may need to adjust
+        //    after extensive testing for typical use cases in the next release.
+        dataPath.getFileSystem(conf).getContentSummary(dataPath).getLength * 3
```

We skip index when "read index file need more time than read data file".

So, I think all your code is based on 1 assumption:
time of reading uncompressed data
is equal to
time of reading same but compressed date + time of decompressing data

Then, why we choose to compress it?

---

Actually I don't have this kind of assumption. IMHO, this cost policy is to tentatively avoid the case that index time is greater than full scan time, because the huge index size may increase the total time due to the cost of high disk IO. The advantage of compression is the reduction of both disk IO and memory consumption in the cost of additional computing to do the decompression. It's worthy doing that. The issue #541 is one example.  Actually, if most of the uncompressed data is remote, IMHO, the time of reading these data is greater than the time of reading compressed and then decompression. Thanks.

---

then why you compare index file with compressed file's uncompressed size?

---

Actually I described the motivation in the code. The index file size is uncompressed. The oap data file is compressed. The default comparison may be misleading to the cases which will use full scan although it may benefit from index. That's unfair from my point of view. You can get more background from the discussion in the issue #509.